Inner flow parallel execution support

Baraujo25 opened the pull request.

Baraujo25 committed:
Execute ParallelPacketStep inner flows in different threads

rabelenda commented on src/main/java/us/abstracta/wiresham/ConnectionFlowDriver.java:
```diff
@@ -33,39 +38,88 @@ public void run() {
-        closeFlowConnections();
-      } catch (IOException e) {
-        LOG.error("Problem while releasing sockets", e);
+        if (parallelStepsExecutor.awaitTermination(PARALLEL_THREADS_TIMEOUT,
```
we are imposing a timeout regardless of the number of steps to the total execution time of sub flows. This sounds fragile and require constant tuning of the timeout for users.

rabelenda replied:
Maybe, to avoid leaving connections open, and not just for parallel execution, we should put some timeout in general of flows activity? Something like: all connections of a flow will be closed after X time of no received packet? Another thing you can do is not put a timeout, and in this case, collect all Futures from submit, iterate over them doing a get, and then shutdown the thread pool.

Baraujo25 replied:
Any idea to face the issue?
 Nowadays when client does not complete the whole flow we are up listening indefinitely. How is that different from this scenario?

EDIT: when writing this comment your second comment wasn't loaded yet IGNORE

Baraujo25 replied:
I took the second idea. The one using Futures get. Since I believe it keeps aligned with previous Wiresham versions idea. The idea of listening indefinitely until a connection is made.

Baraujo25 committed:
Improve join in Parallel threads

Baraujo25 committed:
Add tests for Parallel flow execution

rabelenda commented on src/main/java/us/abstracta/wiresham/ConnectionFlowDriver.java:
```diff
@@ -33,42 +38,92 @@ public void run() {
+      } catch (InterruptedException e) {
+        LOG.error("Parallel steps where interrupted", e);
+      } catch (ExecutionException ignored) {
+        //Should not occur since we are catching all exceptions inside task
```
maybe throw it as runtime? since is not expected, or log it as trace or debug?

rabelenda commented on src/main/java/us/abstracta/wiresham/ConnectionFlowDriver.java:
```diff
@@ -33,42 +38,92 @@ public void run() {
+      try {
+        future.get();
+      } catch (InterruptedException e) {
+        LOG.error("Parallel steps where interrupted", e);
```
Have you tested this? Is it ok to throw an error? I mean, user sends ctrl+c and log error? maybe a warning?, but something important is the amount of logging you generate for one event as well.

rabelenda commented on src/main/java/us/abstracta/wiresham/Flow.java:
```diff
@@ -272,8 +274,16 @@ public int getPortCount() {
+          }
+          return null;
+        })
+        .filter(Objects::nonNull)
```
you can avoid this returning Collections.empty() instead. In general is a good practice for collections to return their empty version instead of returning null.

rabelenda commented on src/main/java/us/abstracta/wiresham/ParallelPacketStep.java:
```diff
@@ -0,0 +1,40 @@
+  public List<Integer> getPorts() {
+    return parallelSteps.stream()
+        .flatMap(Collection::stream)
+        .filter(p -> p instanceof SendPacketStep)
```
why only send packets?

Baraujo25 committed:
Add deserialization of parallel steps

Baraujo25 committed:
Add test for deserialization of parallel steps

Baraujo25 committed:
Properly log unexpected exception when waiting for parallel threads

Baraujo25 committed:
Collect ports form Send and Receive packets

rabelenda commented on src/main/java/us/abstracta/wiresham/Flow.java:
```diff
@@ -45,6 +48,7 @@ public class Flow {
       .put("!server", SendPacketStep.class)
       .put("!client", ReceivePacketStep.class)
       .put("!include", IncludePacketStep.class)
+      .put("!parallel:", ParallelPacketStep.class)
```
is it expected to include colons here?

Baraujo25 replied:
I thought we agree on the design of the yaml to be something like:
```yaml
- !server { data: 48656C6C6F, port: 23}
- !client { data: 48656C6C6F2C2049276D204A6F686E }
- !parallel:
    -
      - !server { data: 48656C6C6F204A6F686E, port: 24 }
      - !client { data: 48656C6C6F2C2049276D204A6F686E }
      - !server { data: 427965204A6F686E }
    - 
      - !server { data: 48656C6C6F, port: 25}
      - !client { data: 48656C6C6F2C2049276D204A6F686E }
     
```
However, it does not complain the standards we had before. 
Brackets is the standard we used. Therefore yaml will look something like:
```yaml
- !server { data: 48656C6C6F, port: 23}
- !client { data: 48656C6C6F2C2049276D204A6F686E }
- !parallel {
    -
      - !server { data: 48656C6C6F204A6F686E, port: 24 }
      - !client { data: 48656C6C6F2C2049276D204A6F686E }
      - !server { data: 427965204A6F686E }
    - 
      - !server { data: 48656C6C6F, port: 25}
      - !client { data: 48656C6C6F2C2049276D204A6F686E }
   }
   ```
That looks aligned with previous yaml syntax. What do you think?

rabelendameli replied:
I wonder now if the first alternative "aligns" with yaml. Would need to review. I initially thought that both alternatives where mainly the same, but it may not be the case. Let's research a little about that.

rabelenda commented on src/main/java/us/abstracta/wiresham/Flow.java:
```diff
@@ -272,8 +276,15 @@ public int getPortCount() {
-        .filter(p -> p instanceof SendPacketStep)
-        .map(PacketStep::getPort)
+        .map(packetStep -> {
+          if (packetStep instanceof SendPacketStep) {
```
why sendPackets? What about receive packets? what about include packets? Maybe just a getPorts in any packet and in single packet scenarios returns an empty or singleton list?

rabelenda commented on src/main/java/us/abstracta/wiresham/Flow.java:
```diff
@@ -285,6 +296,14 @@ public static class FlowConstructor extends Constructor {
 
     @Override
     protected Object constructObject(Node node) {
+      if (node.getTag().getValue().equals("!parallel:")) {
```
again collons, is this ok?

rabelenda replied:
Maybe this was what was messing with lists and required a custom constructor?

Baraujo25 replied:
I left the new design of the yaml in above comment. 
Regarding your second comment. No, it was not the cause of the issue. Since my first approach for the solution was using brackets as we did for `!server`, `!client` and `!include` and the issue persists. Custom constructor is needed due to the lack of parameters presented in the yaml.
The deserialization process require one parameter if the object to be constructed has a constructor with one parameter. The Yaml api requires the explicit parameter which could be something like `!parallel { parallelSteps: ...}` or even a list as we previously saw (`!parallel { - ...}`) The reason why `-` works to deserialize is based on the fact that the `-` is parsed as a SequenceNode which at the end does nothing else that provide the list of lists. It's an unnecessary node from my pov, but understandable from snake-yaml api side. 

For these scenarios snake-yaml could provide a simple api that won't require the specification of a parameter when dealing with a class that only has one field to deserialize.

rabelenda commented on src/main/java/us/abstracta/wiresham/ParallelPacketStep.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class ParallelPacketStep extends PacketStep {
```
Now looking at it, PacketStep should probably be a sub class of FlowStep. Because semantically parallelStep is not a "packet" step, but a multi packet step (as is includeStep)?

Baraujo25 committed:
Re-organize steps hierarchy adding FlowStep

FlowStep was added in order to organize hierarchy in terms of executable steps and collection steps.
Executable steps are going to be PacketSteps which still be a FlowStep
Collection steps are going to be ParallelStep and Include steps. Previously mentioned steps are not executable.

Baraujo25 committed:
Remove semicolon in Parallel tag

Baraujo25 committed:
Add mvn surefire plugin to run test

Baraujo25 committed:
Fix Reload and Tcp Service test

rabelenda commented on src/main/java/us/abstracta/wiresham/ConnectionFlowDriver.java:
```diff
@@ -29,46 +34,97 @@ public ConnectionFlowDriver(FlowConnectionProvider connectionProvider,
+      throws ExecutionException, InterruptedException, IOException {
+    while (!flowSteps.isEmpty()) {
+      FlowStep step = flowSteps.poll();
+      if (step instanceof ParallelPacketStep) {
```
are we missing renaming ParallelPacketStep to ParallelFlowStep?

rabelenda commented on src/main/java/us/abstracta/wiresham/ConnectionFlowDriver.java:
```diff
@@ -29,46 +34,97 @@ public ConnectionFlowDriver(FlowConnectionProvider connectionProvider,
+        processParallelSteps((ParallelPacketStep) step, previousPort, connectionProvider);
+        continue;
+      }
+      PacketStep packetStep = (PacketStep) step;
```
what about include steps? is it previously solved?

Baraujo25 replied:
Yes. Include steps are already flatted as flowSteps. They could either be parallel, receive and send.

Baraujo25 committed:
Rename clases that where impacted for new FlowStep hierarchy

rabelenda commented on src/main/java/us/abstracta/wiresham/ConnectionFlowDriver.java:
```diff
@@ -29,46 +34,97 @@ public ConnectionFlowDriver(FlowConnectionProvider connectionProvider,
+
+  private static void processParallelSteps(ParallelPacketStep step, int previousPort,
+      FlowConnectionProvider connectionProvider) {
+    List<List<FlowStep>> parallelSteps = step.getParallelSteps();
```
we could rename parallelSteps to forks/branches, might it be less confusing? Keep it or leave it.

Baraujo25 replied:
I'll take this comment. Waaay more descriptive! Thanks 💪

Baraujo25 committed:
Make fields of more mnemonic for parallel flow step

Baraujo25 committed:
Fix Reload Service test
